Re-write the LCB VM

livecodefraser opened the pull request:
Like https://github.com/livecode/livecode/pull/4457 but with bugfixes.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/livecode/livecode/4464)

<!-- Reviewable:end -->

runrevmark committed:
[[ Foundation ]] Add an initialising constructor to MCAutoPointer

runrevmark committed:
[[ Foundation ]] Add new forms of MCError construction

This patch adds new constructors for MCError, allowing varargs or va_list forms.

runrevmark committed:
[[ Foundation ]] Add short-circuit check to MCTypeInfoConforms

This patch checks to see if the the source and target typeinfos are
the same before doing any further analysis.

runrevmark committed:
[[ Script ]] Rework LCB VM implementation

This patch contains a significant rework of the LCB VM implementation.

Most notably, individual bytecodes are now described as individual structs,
and the main operations of description (for building), validation and
execution for each bytecode is within the struct.

runrevmark committed:
[[ Script ]] Tweaks to iterators

In order to ensure iterators still work in the new VM, some tweaks were
necessary to the standard library implementations.

livecodefraser committed:
Fix a crash when trying to import unbridgeable foreign types

livecodefraser committed:
Remove a too-strict assertion from MCTypeInfoConforms

The VM needs to be able to check for conformance at times when it
doesn't have access to the type names.

livecodefraser committed:
Declare types to conform if they have the same bridge type

livecodefraser committed:
Make it an error to have ambiguous overloads in handler groups

livecodefraser committed:
Add "NothingIs[Not]EqualToNothing" overloads for "is [not]"

This is required to resolve ambiguity when both sides are undefined.

livecodefraser committed:
Remove "is x among y" syntax for codeunits

This overload was ambiguous and could never be called because the chars
form was always chosen instead.

livecodefraser committed:
Mark all LCB iterator syntax as having optional iteration vars

Otherwise, it isn't possible to use uninitialised vars for the
iteration variable.

livecodefraser committed:
Fix copying foreign values to foreign-typed slots

The case where both were the same (foreign) type wasn't handled.

livecodefraser committed:
Fix passing of undefined variables of unbridgeable foreign types

The case of unbridgeable foreign types (like pointer) wasn't handled so
optional forms of these types weren't usable as parameters as they'd
fail due to the lack of bridging from undefined -> pointer.

livecodefraser committed:
Allow foreign -> foreign type conversions

Auto-converion of foreign types now happens if they are the same type
or if they share a bridging type. Without this, there is no way to
convert between different types of C integers.

livecodefraser committed:
Ensure LCB handler calls use the correct register mapping

This fixes inout/out parameters misbehaving in some circumstances.

livecodefraser committed:
Update the canvas module to handle more number types in lists

It now handles any foreign type that bridges to MCNumberRef, not just
NumberRefs themselves. This fixes problems when lists contain foreign
number types (like C floats or integers).

livecodefraser committed:
The ZString LCB tests are now expected to pass

The failure to bridge is properly detected.

livecodefraser commented:
Bits of this change are likely to be temporary. In particular, the changes to the way type bridging works should be obsoleted when the type system is re-done.

Things still to be done with this PR:
- [x] Get all the widgets in the IDE working again (browser in particular)
- [x] Update the compiler to emit slightly different bytecode for repeat loops
- [x] Fix the other bugs I haven't found yet

livecodefraser committed:
Mark the context pointer for libbrowser callbacks as optional

Depending on the browser implementation, this pointer can be null.

livecodefraser committed:
Mark the native display pointer for widgets as optional

The parameter is only used on X11 and is null everywhere else.

livecodefraser committed:
Fix an uninitialised Rectangle in paletteactions

There isn't yet the capability to have default initialisers for custom
types.

livecodefraser committed:
Use the resolved return type when generating FFI CFIs

It needs to be checked against the special value kMCNullTypeInfo.

livecodefraser committed:
Add more assertions that typeinfos are passed in named form

livecodefraser committed:
Revert "Remove a too-strict assertion from MCTypeInfoConforms"

This reverts commit d581be6098cce91a5a41f8c15cf0c2c11c24b891.

The underlying cause has been fixed so this assertion can be reinstated.

livecodefraser committed:
Change UnboxingConvert to always bridge

Not bridging here complicates implementations of various things as they
have to do the conversions themselves.

livecodefraser committed:
Use the named type when creating foreign values

livecodefraser committed:
Allow bridgeable optional foreign types to import correctly

livecodefraser committed:
Auto-bridge foreign types when exporting to LCS-world

livecodefraser commented:
Updating the compiler to emit different bytecode for repeat loops will be done separately as 'repeat' needs to be refactored a bit anyway.

peter-b commented:
@livecodefraser This mostly looks good to me.  However, I think it's necessary to ensure that the "`repeat` over empty list into strongly-typed `variable`" pattern doesn't cause a run-time error before merging this PR.

livecodefraser committed:
Merge branch 'develop' of https://github.com/runrev/livecode into vm-refactor

Conflicts:
	libscript/src/script-instance.cpp

livecodefraser committed:
Merge branch 'lcb-repeat_temp_reg' into vm-refactor

Conflicts:
	libfoundation/include/foundation-auto.h

livecodefraser committed:
Add a describe implementation for ScriptObjects

livecodefraser commented:
Includes https://github.com/livecode/livecode/pull/4596

livecodefraser committed:
Revert "Mark all LCB iterator syntax as having optional iteration vars"

This reverts commit 19186f91f8d8e97e975a1e5b2ec2485162bffb56.

livecodefraser committed:
Revert "[[ Script ]] Tweaks to iterators"

This reverts commit eab38131ec206d05ae6f0b503032d41bfd68bdfb.

Conflicts:
	libscript/src/codeunit.lcb

livecodefraser committed:
Add a test ensuring iterators use a temporary reg

livecodefraser committed:
[[ Bugfix 18350 ]] Add bugfix note

peter-b commented:
@livecode-vulcan review ok f813332

livecode-vulcan commented:
💙 review by @peter-b ok f813332434c9131e0044716d1fbec682f6e88707

livecode-vulcan commented:
😞 test failure f813332434c9131e0044716d1fbec682f6e88707
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/417)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/422)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/417)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/417)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/423)
- try-community-win-x86: [failure](https://vulcan.livecode.com/builders/try-community-win-x86/builds/417)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/417)

peter-b commented:
@livecodefraser Looks like there's an unrecognised pragma making the MSVC 2010 compiler sad.

livecodefraser committed:
Guard some GCC-specific warning pragmas in libscript

livecodefraser committed:
Add magic to foundation.h to provide alignof everywhere

C++11 added alignof(x) to return the alignment of a type. Not all of our
compilers support this but they do all have equivalents. Lots of ifdef
magic has been added to foundation.h to support this.

peter-b commented on libscript/src/script-execute.cpp:
```diff
@@ -116,7 +116,7 @@ class MCScriptForeignInvocation
 	{
 		if (m_argument_count == kMaxArguments ||
 			!Allocate(sizeof(void *),
-					  __alignof__(void *),
```
@livecodefraser Do you think it might be better to put a `#define __alignof__ sizeof` in `foundation.h` (protected by appropriate platform-specific `#ifdef`ery)?  I think this would make it easier to fix properly later on.

peter-b commented:
@livecode-vulcan review ok f3b9012

livecode-vulcan commented:
💙 review by @peter-b ok f3b90120efe0431c11b37fa8e57346fadb4c8404

livecode-vulcan commented:
😎 test success f3b90120efe0431c11b37fa8e57346fadb4c8404
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/446)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/451)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/446)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/446)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/452)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/446)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/446)
